lib/SILOptimizer/IPO/CrossModuleOptimization.cpp

```diff
@@ -835,14 +835,14 @@ void CrossModuleOptimization::makeDeclUsableFromInline(ValueDecl *decl) {
 
   if (!isPackageOrPublic(decl->getFormalAccess()) &&
       !decl->isUsableFromInline()) {
-    // Mark the nominal type as "usableFromInline".
-    // TODO: find a way to do this without modifying the AST. The AST should be
-    // immutable at this point.
-    auto &ctx = decl->getASTContext();
-    auto *attr = new (ctx) UsableFromInlineAttr(/*implicit=*/true);
-    decl->getAttrs().add(attr);
-
     if (everything) {
+      // Mark the nominal type as "usableFromInline".
```

Comment:
This will degrade performance in aggressive mode significantly, e.g. for the SwiftCompilerSource, which we build in aggressive CMO mode.
You should also allow adding the usableFromInline attribute in aggressive mode.